sycl/doc/extensions/proposed/sycl_ext_oneapi_if_device.asciidoc

```diff
@@ -0,0 +1,171 @@
+
+Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
+of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
+permission by Khronos.
+
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/intel/llvm/issues
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 6 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+
+== Overview
+
+This extension adds a way to conditionally select code based on whether the
+caller runs on a device or runs on the host.  This allows an application to
+call some common function from either host or device, and then select one path
+or another for each case.  The `+__SYCL_DEVICE_ONLY__+` macro from the core
+SYCL specification provides this same ability, but the new API in this
+extension provides a solution that works even in a 1-pass implementation of
+SYCL.  In such an implementation, the preprocessor runs only once, so there is
+no way to define the `+__SYCL_DEVICE_ONLY__+` macro differently for the host
+and device compilations.
+
+The goal of this extension is to provide a better alternative to the
+`+__SYCL_DEVICE_ONLY__+` macro, adopt it into a future version of the SYCL
+specification, and deprecate usage of that macro.
+
+The APIs added in this extension are designed in the same style as the APIs in
+link:sycl_ext_oneapi_if_device_has.asciidoc[sycl_ext_oneapi_if_device_has] in
+order to provide a consistent style for conditionally selecting code based on
+where it runs.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_IF_DEVICE` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== New `if_device` and `if_host` free functions
+
+This extension adds two new free functions which may be called from either
+device or host code.
+
+```
+namespace sycl::ext::oneapi::experimental {
+
+template<typename T>
+/* unspecified */ if_device(T fn);
```

Comment:
Why is the callable passed by value? std (e.g. apply) uses perfect forwarding for APIs taking a callable. It would matter if the capture of the callable contains a move only type. In that cases pass by value isn't possible. Sorry didn't notice this on previous reviews.

Reply:
I think I was following `queue::submit` as a model, which also passes the callable by value.  Is there a reason we would use pass-by-value for `queue::submit` and perfect forwarding here?  Or, is `queue::submit` just a bad model to follow?

Reply:
I don't think there is any advantage for `queue::submit` to use pass by value. And therefore not a good model to follow.

Reply:
Thanks, @rolandschulz.  I addressed this in 9cbc078b7e49b5cadb1b461190f60e30a31ef850.